cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -0,0 +1,1100 @@
+
+  // Nominal case -> walk a single bucket
+  return impl_->Walk(select, base_path.bucket, base_path.key, out);
+}
+
+Status S3FileSystem::CreateDir(const std::string& s, bool recursive) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+
+  if (path.key.empty()) {
+    // Create bucket
+    return impl_->CreateBucket(path.bucket);
+  }
+
+  // Create object
+  if (recursive) {
+    // Ensure bucket exists
+    RETURN_NOT_OK(impl_->CreateBucket(path.bucket));
+    // Ensure that all parents exist, then the directory itself
+    std::string parent_key;
+    for (auto it = path.key_parts.begin(); it != path.key_parts.end(); ++it) {
+      parent_key += *it;
+      parent_key += kSep;
+      RETURN_NOT_OK(impl_->CreateEmptyObject(path.bucket, parent_key));
+    }
+    return Status::OK();
+  } else {
+    // Check parent dir exists
+    if (path.has_parent()) {
+      S3Path parent_path = path.parent();
+      bool exists;
+      RETURN_NOT_OK(impl_->IsNonEmptyDirectory(parent_path, &exists));
+      if (!exists) {
+        RETURN_NOT_OK(impl_->IsEmptyDirectory(parent_path, &exists));
+      }
+      if (!exists) {
+        return Status::IOError("Cannot create directory '", path.full_path,
+                               "': parent directory does not exist");
+      }
+    }
+
+    // XXX Should we check that no non-directory entry exists?
+    // Minio does it for us, not sure about other S3 implementations.
+    return impl_->CreateEmptyDir(path.bucket, path.key);
+  }
+}
+
+Status S3FileSystem::DeleteDir(const std::string& s) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+
+  if (path.empty()) {
+    return Status::NotImplemented("Cannot delete all S3 buckets");
+  }
+  RETURN_NOT_OK(impl_->DeleteDir(path.bucket, path.key));
+  if (path.key.empty()) {
+    // Also delete bucket
+    S3Model::DeleteBucketRequest req;
+    req.SetBucket(ToAwsString(path.bucket));
+    return OutcomeToStatus(impl_->client_->DeleteBucket(req));
+  }
+  // Parent may be implicitly deleted if it became empty, recreate it
+  return impl_->EnsureParentExists(path);
+}
+
+Status S3FileSystem::DeleteFile(const std::string& s) {
+  S3Path path;
+  RETURN_NOT_OK(S3Path::FromString(s, &path));
+  RETURN_NOT_OK(ValidateFilePath(path));
+
+  // Check the object exists
+  S3Model::HeadObjectRequest req;
+  req.SetBucket(ToAwsString(path.bucket));
+  req.SetKey(ToAwsString(path.key));
+
+  auto outcome = impl_->client_->HeadObject(req);
+  if (!outcome.IsSuccess()) {
+    if (IsNotFound(outcome.GetError())) {
+      return PathNotFound(path);
+    } else {
+      return ErrorToStatus(outcome.GetError());
+    }
+  }
+  // Object found, delete it
+  RETURN_NOT_OK(impl_->DeleteObject(path.bucket, path.key));
+  // Parent may be implicitly deleted if it became empty, recreate it
+  return impl_->EnsureParentExists(path);
+}
+
+Status S3FileSystem::Move(const std::string& src, const std::string& dest) {
+  // XXX Cannot really move directories
```

Comment:
Let's open JIRA to implement it as a follow up

Reply:
What I mean is: S3 doesn't allow you to "move" (or rename) anything directly. To move a directory, you have to copy all its individual contents (objects), then delete the original ones. It may be quite costly (including financially), unlike the usual expectation for a move operation.